File: test/cpp/case35_alias_arg.cpp
#include "cpp_module/py_runtime.h"

// 同一インスタンス共有のテスト。
// 参照セマンティクスが壊れると a,b の値が一致しない。


struct Box : public PyObj {
    int64 v;
    
    Box(int64 v) {
        this->v = v;
    }
};

void bump(const rc<Box>& x) {
    x->v++;
}

void run_case35_alias_arg() {
    rc<Box> a = rc_new<Box>(1);
    rc<Box> b = a;
    bump(b);
    py_print(a->v);
    py_print(b->v);
}

int main(int argc, char** argv) {
    pytra_configure_from_argv(argc, argv);
    run_case35_alias_arg();
    return 0;
}

File: test/py/case35_alias_arg.py
# 同一インスタンス共有のテスト。
# 参照セマンティクスが壊れると a,b の値が一致しない。


class Box:
    def __init__(self, v: int) -> None:
        self.v = v


def bump(x: Box) -> None:
    x.v += 1


def run_case35_alias_arg() -> None:
    a = Box(1)
    b = a
    bump(b)
    print(a.v)
    print(b.v)


if __name__ == "__main__":
    run_case35_alias_arg()
